File: backend-java/src/main/java/com/example/modules/payments/internal/PaymentService.java
// Business logic — hidden from the outside world.
package com.example.modules.payments.internal;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class PaymentService {
    public static class Payment {
        public String paymentId, orderId, status;
        public long amountCents;
        public Payment(String paymentId, String orderId, String status, long amountCents) {
            this.paymentId = paymentId; this.orderId = orderId; this.status = status; this.amountCents = amountCents;
        }
    }

    private static final Map<String, Payment> STORE = new ConcurrentHashMap<>();

    public Payment authorize(String paymentId, String orderId, long amountCents) {
        Payment p = new Payment(paymentId, orderId, "AUTHORIZED", amountCents);
        STORE.put(paymentId, p);
        return p;
    }

    public Optional<Payment> capture(String paymentId) {
        Payment p = STORE.get(paymentId);
        if (p == null || !"AUTHORIZED".equals(p.status)) return Optional.empty();
        p.status = "CAPTURED";
        return Optional.of(p);
    }

    public Optional<Payment> refund(String paymentId) {
        Payment p = STORE.get(paymentId);
        if (p == null || !"CAPTURED".equals(p.status)) return Optional.empty();
        p.status = "REFUNDED";
        return Optional.of(p);
    }
}

File: backend-scala/src/main/scala/com/example/modules/payments/internal/PaymentService.scala
// Business logic — hidden from the outside world.
package com.example.modules.payments.internal

import java.util.concurrent.ConcurrentHashMap

final case class Payment(paymentId: String, orderId: String, var status: String, amountCents: Long)

class PaymentService:
  private val store = new ConcurrentHashMap[String, Payment]()

  def authorize(paymentId: String, orderId: String, amountCents: Long): Payment =
    val p = Payment(paymentId, orderId, "AUTHORIZED", amountCents)
    store.put(paymentId, p)
    p

  def capture(paymentId: String): Option[Payment] =
    Option(store.get(paymentId)).filter(_.status == "AUTHORIZED").map { p => p.status = "CAPTURED"; p }

  def refund(paymentId: String): Option[Payment] =
    Option(store.get(paymentId)).filter(_.status == "CAPTURED").map { p => p.status = "REFUNDED"; p }
